### DISCUSSION TITLE
Concurrent program analysis using state machines

### DISCUSSION THREAD

**Discussion by dmlloyd** [qbicc/qbicc#1109] [Design discussions]
**TL;DR:** We can make concurrent program analysis simpler, safer, more robust, and more extensible by managing the analysis state for program elements as state machines.

## Rationale

Several of our areas of analysis involve accumulating facts about various program elements, and then acting on those facts when various conditions are met. These disparate facts can, in many cases, be individually discovered in multiple possible orders, and only when certain sets of facts are known about an element should action be taken. Furthermore, the action usually must be executed at most one time. And to make matters even worse, the number of real states goes up exponentially with the number of individual facts; for example, 10 distinct binary (present/not present) facts about some element can result in as many as 2<sup>10</sup> = 1024 possible states.

At present, management of interrelated states is done in an ad-hoc manner.  Different but related facts are stored in separate structures. State transitions may be entered from many points, raising the possibility of inconsistent handling between two otherwise similar transitions. Since each transition is handled separately, problems must be solved as they are encountered with only sporadic correctness assertions.

This all adds up to significant complexity and room for error. Since it is likely that this style of analysis is going to be normal for this project, a more general solution is called for.

## Characteristics

The proposed solution has the following characteristics:

* Facts: a fact is defined as a binary (present/not present) state which establishes some deterministic understanding of a program element.
* Centralization of state: all facts for a given kind of program element should, whenever possible, be collected at one point.
* Fact discovery is idempotent by nature, so that registering the discovery of a fact that was already discovered for the same program element should have no additional effect.
* Open access to fact reporting: the API for reporting newly discovered facts should be simple and uniform; for example, when tracking field accesses, the API could allow for any caller to register facts for a particular field element, such as `FIELD_READ` or `FIELD_WRITTEN`.
* Identification of transition conditions: a mechanism should be provided whereby transition conditions can be clearly codified and registered (for example, when facts `A`, `B` and (`C` or `D`) are met (in any order), action `X` shall be performed exactly once).
* Composability of actions: to give an example, a fact of `FOUND_ON_HEAP` might be discovered for a class by a serialization-related subsystem; consequently, all of that class's fields might be considered to have been written. An action could be registered on that fact to discover the `FIELD_WRITTEN` fact for all of that class's fields, and any other subsystem that is interested in whether a field was written to may act on that information without the necessity of the original subsystem's knowledge.
* Immunity to recursion circularity problems: since fact discovery is idempotent, chained actions can never be circularly executed even in the presence of actions which might otherwise have circular causality, though the order in which the actions execute may be arbitrary in any case.
* Upper bounds: any number of facts or actions per program element should be supportable.

## API

The fact management API for a given kind of program element would offer a means for registering fact discovery on a given program element of that kind, accepting the element and the discovered fact(s) for that element.  It would also have a means for registering an action to be taken, which would be given the condition to match and the action to take when the condition is met.

It is possible that several different combinations of facts might suffice to cause an action to be taken. The API should provide for this by offering conditions which are satisfied by the presence of single fact, the presence of all of a set of sub-conditions, or the presence of any of a set of sub-conditions.

For example, I might register an action as follows:

```java
    fieldStates.registerAction(Fact.any(FieldAccessFact.FIELD_READ, FieldAccessFact.FIELD_WRITTEN), fe -> System.out.printf("Field %s was accessed!%n", fe.getName());
```

I might discover a fact like this:

```java
    fieldStates.discover(thisField, FieldAccessFact.FIELD_WRITTEN);
```

Or I could discover multiple facts at once:

```java
    fieldStates.discover(thisField, FieldAccessFact.FIELD_READ, FieldAccessFact.FIELD_WRITTEN);
```

either of which *may* in turn trigger the previously registered action if it hasn't already been triggered.

## Phases

As with all of our analyses, per #692 and related issues, the program element states should be cycled on each phase, so that every phase has access to the prior phase's full analysis results, even while discovering facts anew on the current phase.

## Limitations

This solution only envisions support for binary facts. Other kinds of fact may be interesting; for example, we might track how many times a method is invoked, or how many instances of a given type might be present, so that we could potentially take some action when a threshold is reached.

It would be difficult to support registration of actions after facts have been discovered; I would not plan to support this.

It would be useful to provide some kind of context such as a `Location` which can be used in diagnostics which might be produced by an action.

## RFC

This is a design proposal and as such it is open to discussion and feedback. Please let me know if you have any thoughts. My plan is to apply this concept for fields to the draft PR #1059, and work on expanding it to other areas if that is successful.
